classes/class.ilPCExternalContent.php

```diff
@@ -6,110 +6,94 @@
     {
+        global $DIC;
+        $this->tree = $DIC->repositoryTree();
+        $this->http = $DIC->http();
+        $this->refinery = $DIC->refinery();
         $this->settings = new ilExternalContentSettings($settings_id);
         $this->plugin = $plugin;
     }
 
     /**
      * Get the settings object
-     * @return ilExternalContentSettings
      */
-    public function getSettings()
+    public function getSettings(): ilExternalContentSettings
     {
         return $this->settings;
     }
 
     /**
      * Get the object id of the parent object (learning module, content page)
-     * @return int
      */
-    public function getId()
+    public function getId(): int
     {
         return $this->plugin->getParentId();
     }
 
-
     /**
      * Get the reference id of the parent object (learning module, content page)
-     * @return int
      */
-    public function getRefId()
+    public function getRefId(): int
     {
-        return $_GET['ref_id'];
+        return $this->http->wrapper()->query()->retrieve('ref_id', $this->refinery->kindlyTo()->int());
     }
 
     /**
      * Get the title of the parent object (learning module, content page)
-     * @return string
      */
-    public function getTitle()
+    public function getTitle(): string
     {
         return ilObject::_lookupTitle($this->getId());
     }
 
     /**
      * Get the description of the parent object (learning module, content page)
-     * @return string
      */
-    public function getDescription()
+    public function getDescription(): string
     {
         return ilObject::_lookupDescription($this->getId());
     }
 
     /**
      * Get the higher context (course or group of the parent object)
-     * @return array
+     * @return array{id: int, title: string, type: string}
      */
-    public function getContext()
+    public function getContext(): array
     {
         /** @see ilObjExternalContent::getContext() */
         $valid_types = array('crs', 'grp', 'cat', 'root');
-        global $DIC;
-        $tree = $DIC->repositoryTree();
-        if (!isset($this->context)) {
-            $this->context = array();
-            $path = array_reverse($tree->getPathFull($this->getRefId()));
-            foreach ($path as $key => $row)
-            {
-                if (in_array($row['type'], $valid_types))
-                {
-                    if (in_array($row['type'], array('cat', 'root')) && !empty($this->context))
+        if ($this->context === null) {
+            $this->context = [];
+            $path = array_reverse($this->tree->getPathFull($this->getRefId()));
+            foreach ($path as $key => $row) {
+                if (in_array($row['type'], $valid_types)) {
+                    if (in_array($row['type'], array('cat', 'root')) && !empty($this->context)) {
                         break;
-
+                    }
                     $this->context['id'] = $row['child'];
```

Comment:
Maybe we should ensure `child` is an integer. This seems to be not guaranteed by `\ilTree::fetchNodeData`, but AFAIK real integers are used since PHP 8.1. Nevertheless, I suggest making this explicit.

```suggestion
                    $this->context['id'] = (int) $row['child'];
```